src/compiler/checker.ts

```diff
@@ -24690,6 +24690,19 @@ namespace ts {
                 // If the current type is a union type, remove all constituents that couldn't be instances of
                 // the candidate type. If one or more constituents remain, return a union of those.
                 if (type.flags & TypeFlags.Union) {
+                    if(candidate.flags & TypeFlags.Union){
+                        const assignableType = mapType(type, type => {
+                            if (!!(type.flags & TypeFlags.String) || !!(type.flags & TypeFlags.Number)) {

```

Comment:
You definitely don't just want to filter to `String` or `Number` types; there's nothing particularly magical about them - what's wrong with applying this logic to all types?

Reply:
I certainly had no problem applying it to all types.